node/modules/storageminer.go

```diff
@@ -357,76 +356,187 @@ func SectorStorage(mctx helpers.MetricsCtx, lc fx.Lifecycle, lstor *paths.Local,
 	return sst, nil
 }
 
-func F3Participation(mctx helpers.MetricsCtx, lc fx.Lifecycle, api v1api.FullNode, minerAddress dtypes.MinerAddress) error {
-	ctx := helpers.LifecycleCtx(mctx, lc)
-	b := &backoff.Backoff{
-		Min:    1 * time.Second,
-		Max:    1 * time.Minute,
-		Factor: 1.5,
-		Jitter: false,
+type f3Participator struct {
+	node                     v1api.FullNode
+	participant              address.Address
+	backoff                  *backoff.Backoff
+	timer                    *time.Timer
+	maxCheckProgressAttempts int
+	previousTicket           api.F3ParticipationTicket
+	checkProgressInterval    time.Duration
+	leaseTerm                uint64
+}
+
+func newF3Participator(node v1api.FullNode, participant dtypes.MinerAddress, backoff *backoff.Backoff, maxCheckProgress int, checkProgressInterval time.Duration, leaseTerm uint64) *f3Participator {
+	return &f3Participator{
+		node:                     node,
+		participant:              address.Address(participant),
+		backoff:                  backoff,
+		timer:                    time.NewTimer(0),
+		maxCheckProgressAttempts: maxCheckProgress,
+		checkProgressInterval:    checkProgressInterval,
+		leaseTerm:                leaseTerm,
 	}
-	go func() {
-		timer := time.NewTimer(0)
-		defer timer.Stop()
+}
 
-		if !timer.Stop() {
-			<-timer.C
+func (p *f3Participator) participate(ctx context.Context) error {
+	for ctx.Err() == nil {
+		if ticket, err := p.tryGetF3ParticipationTicket(ctx); err != nil {
+			return err
+		} else if lease, participating, err := p.tryF3Participate(ctx, ticket); err != nil {
+			return err
+		} else if !participating {
+			continue
+		} else if err := p.awaitLeaseExpiry(ctx, lease); err != nil {
+			return err
 		}
+		log.Info("Restarting F3 participation")
+	}
+	return ctx.Err()
+}
 
-		leaseTime := 120 * time.Second
-		// start with some time in the past
-		oldLease := time.Now().Add(-24 * time.Hour)
-
-		for ctx.Err() == nil {
-			newLease := time.Now().Add(leaseTime)
+func (p *f3Participator) tryGetF3ParticipationTicket(ctx context.Context) (api.F3ParticipationTicket, error) {
+	p.backoff.Reset()
+	for ctx.Err() == nil {
+		switch ticket, err := p.node.F3GetOrRenewParticipationTicket(ctx, p.participant, p.previousTicket, p.leaseTerm); {
+		case ctx.Err() != nil:
+			return api.F3ParticipationTicket{}, ctx.Err()
+		case errors.Is(err, api.ErrF3Disabled):
+			log.Errorw("Cannot participate in F3 as it is disabled.", "err", err)
+			return api.F3ParticipationTicket{}, xerrors.Errorf("acquiring F3 participation ticket: %w", err)
```

Comment:
Won't this be the case while bootstrapping, switching networks, etc?

Reply:
I am not sure. 

> bootstrapping
> switching networks

Good point: Do we disable F3 or pause the inner runner? 🤔 

I expect that the F3 is enabled when the miner starts. Does/Should Lotus support runtime reconfiguration at runtime for `F3Enabled` flag?

Looking at [the API implementation](https://github.com/filecoin-project/lotus/blob/3d12ca5c53c5df74bb9daa0ea9be27b283e2469c/node/impl/full/f3.go) the `api.ErrF3Disabled` is returned when `lf3.F3` is nil, which translates to "Not injected by DI" I believe, which means `F3Enabled` must have been false. If Lotus does not support runtime update for this config then I think the execution path for both node and the miner is that should be enabled before the process starts. Which means `lf3.F3` will not be nil prior to bootstrapping or at the time of network switch. Which means `api.ErrF3Disabled` will not be returned. 

Right?

Reply:
Ah, I see. You're right.